Add `workflow_dispatch` support to PR Help Wanted check

BagToad opened the pull request:
This update allows the PR Help Wanted workflow to be triggered manually via `workflow_dispatch` with a specified PR URL. It also adds logic to fetch some PR details using the GitHub CLI for manual runs.


Fixes https://github.com/cli/cli/pull/11179
<!--
  Thank you for contributing to GitHub CLI!
  To reference an open issue, please write this in your description: `Fixes #NUMBER`
-->

BagToad committed:
Add workflow_dispatch support to PR Help Wanted check

This update allows the PR Help Wanted workflow to be triggered manually via workflow_dispatch with a specified PR URL. It adds logic to fetch PR details using the GitHub CLI for manual runs and unifies variable handling for both event types.

andyfeller commented on .github/workflows/pr-help-wanted.yml:
```diff
@@ -2,6 +2,12 @@ name: PR Help Wanted Check
 on:
   pull_request_target:
     types: [opened]
+  workflow_dispatch:
+    inputs:
+      pr_url:
+        description: "Pull Request URL to check"
+        required: true
+        type: string
```
**suggest:** could we change this to `pr_number` and construct the PR url in the step below?

Since we are using a token other than `secrets.GITHUB_TOKEN`, I'm a little worried security wise about what kinds of URLs might be provided.  The thought being that we could form the URL within the workflow before calling the script.

Am I being too paranoid?

BagToad replied:
> Since we are using a token other than `secrets.GITHUB_TOKEN`, I'm a little worried security wise about what kinds of URLs might be provided. The thought being that we could form the URL within the workflow before calling the script.

This is a valid concern when handling untrusted input, but I initially chose to dismiss it because only those users with write access to the repo should be able to run the workflow manually like this.

Also, on the off chance that a non-github URL gets injected here, `gh` will protect us by not sending a `GITHUB_TOKEN` to any host that isn't github.com or a subdomain.

That all said, I am 100% on board with being paranoid and will make the change as suggested 💯 The ergonomics of just putting in a PR number instead is also desirable in my opinion. Thanks! 🥂 

Fixed in 61e9d94033047f02c83eefd5c5572c39d70cd364

andyfeller replied:
Thank you for understanding my concerns! 🙇

To be clear, I wasn't thinking about non-GitHub URLs so much as URLs to PRs outside of `cli/cli`.

andyfeller commented on .github/workflows/pr-help-wanted.yml:
```diff
@@ -15,13 +21,31 @@ jobs:
           PR_AUTHOR_ASSOCIATION: ${{ github.event.pull_request.author_association }}
-        if: "!github.event.pull_request.draft"
+          PR_URL: ${{ github.event.pull_request.html_url || steps.pr-vars-dispatch.outputs.pr_url }}
+        if: github.event.pull_request.draft == 'false' || steps.pr-vars-dispatch.outputs.is_draft == 'false'
```
**thought:** we could move this logic into the script for more consistent testing if we don't expect to run this script in other situations where that logic doesn't make sense.

not the end of the world

BagToad replied:
Done in 8134fb377b4686e4f67aa79160c3b14721f7985d

andyfeller reviewed (commented):
Looks straight forward!  One minor security question / concern I think we can clear up and get shipped quickly

BagToad committed:
Update workflow to use PR number instead of URL

Changed the workflow_dispatch input from 'pr_url' to 'pr_number' and updated the script to construct the PR URL from the number.

babakks commented on .github/workflows/pr-help-wanted.yml:
```diff
@@ -15,13 +21,32 @@ jobs:
 
+      - name: Set PR variables for workflow_dispatch event
+        id: pr-vars-dispatch
+        if: github.event_name == 'workflow_dispatch'
```
Question; shouldn't the `if` value be enclosed by `{{` and `}}`? I mean, like [this](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#example-using-contexts) from the docs.

BagToad replied:
It actually does not need to be when inside an `if` because `if` values are implicitly expressions.

andyfeller replied:
From https://docs.github.com/en/actions/reference/evaluate-expressions-in-workflows-and-actions#about-expressions:

> The exception to this rule is when you are using expressions in an if clause, where, optionally, you can usually omit ${{ and }}. For more information about if conditionals, see [Workflow syntax for GitHub Actions](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idif).

For what its worth, I typically don't include the evaluation brackets for `if` because it is implicit whereas other places need to distinguish between literal and interpreted text.

BagToad replied:
[reference](https://docs.github.com/en/actions/reference/evaluate-expressions-in-workflows-and-actions#about-expressions)

<img width="809" alt="image" src="https://github.com/user-attachments/assets/4e058784-d024-4ecc-965e-eabbefb31076" />

babakks commented on .github/workflows/pr-help-wanted.yml:
```diff
@@ -15,13 +21,32 @@ jobs:
           PR_AUTHOR_ASSOCIATION: ${{ github.event.pull_request.author_association }}
-        if: "!github.event.pull_request.draft"
+          PR_URL: ${{ github.event.pull_request.html_url || steps.pr-vars-dispatch.outputs.pr_url }}
+        if: github.event.pull_request.draft == 'false' || steps.pr-vars-dispatch.outputs.is_draft == 'false'
```
And also here.

BagToad committed:
Move help-wanted check for draft PRs into script

BagToad committed:
Don't prefix URL with `#`

babakks commented on .github/workflows/pr-help-wanted.yml:
```diff
@@ -15,13 +21,32 @@ jobs:
           PR_AUTHOR_ASSOCIATION: ${{ github.event.pull_request.author_association }}
-        if: "!github.event.pull_request.draft"
+          PR_URL: ${{ github.event.pull_request.html_url || steps.pr-vars-dispatch.outputs.pr_url }}
+          PR_IS_DRAFT: ${{ github.event.pull_request.draft == 'false' || steps.pr-vars-dispatch.outputs.is_draft == 'false' }}
```
Now that we're doing it, why not move the draft detection into the script as well?

BagToad replied:
I thought I moved it to the script as much as I could. Could you expand more on what you mean?

BagToad replied:
Do you mean we should not fetch the PR state at all in the workflow file and instead fetch it in the script?

andyfeller replied:
Yeah, I believe that is what Babak is suggesting.  It would simplify the workflow especially with kicking it off manually and needing to retrieve the same information in a different conditional step.

babakks replied:
Yeah, that's exactly my point. The workflow would be simpler that way. It's also easier to read/understand because all is happening in the script, and there's no external context to track down.

BagToad replied:
Thanks folks! Done in b16347c42de18c025c822c9cd4d01cfa8a3e9487

babakks reviewed (commented):
Thanks for the changes! Just a question.

BagToad committed:
Invert draft checking logic

Inverting this logic because anything other than "false" means we should skip it.

BagToad committed:
Move PR draft status check to shell script

The logic for checking if a pull request is a draft has been moved from the GitHub Actions workflow YAML to the check-help-wanted.sh script. This simplifies the workflow file and centralizes the draft status check within the script.

andyfeller commented on .github/workflows/pr-help-wanted.yml:
```diff
@@ -15,13 +21,27 @@ jobs:
+      - name: Set PR variables for workflow_dispatch event
+        id: pr-vars-dispatch
+        if: github.event_name == 'workflow_dispatch'
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          PR_NUMBER: ${{ github.event.inputs.pr_number }}
```
**nit:**  we could remove the `GH_TOKEN` env var, which I suspect was carried over from the other step

```suggestion
        env:
          PR_NUMBER: ${{ github.event.inputs.pr_number }}
```

BagToad replied:
Sorry missed this with the auto merge 👀 I'll fix it

andyfeller commented on .github/workflows/pr-help-wanted.yml:
```diff
@@ -15,13 +21,27 @@ jobs:
       - name: Check for issues without help-wanted label
         env:
           GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+            # These variables are optionally used in the check-help-wanted.sh
+            # script for additional checks; but they are not strictly necessary
+            # for the script to run. This is why we are okay with them being
+            # empty when the event is workflow_dispatch.
```
**praise:** thank you for including the commentary!

I agree; for right now, they aren't necessary.  If / when they are, we can revise to pull the information from the script itself.

andyfeller reviewed (approved):
Looks good to me!
